### PR TITLE
Query: Add a hook to override single projection or predicate

### PR DESCRIPTION
**Pull request by smitpatel**
@maumar @anpete

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 1]
```diff
@@ -210,7 +210,7 @@ public virtual Expression VisitSelect(SelectExpression selectExpression)
                     _relationalCommandBuilder.Append(", ");
                 }
 
-                VisitProjection(selectExpression.Projection);
+                VisitJoin(selectExpression.Projection, VisitProjection);
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: rename VisitJoin to something else the name is confusing a bit

Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: IIRC, VisitJoin is just an existing private helper method.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 2]
```diff
@@ -315,13 +308,26 @@ var newExpression
         }
 
         /// <summary>
-        ///     Visit the projection.
+        ///     Visit a single projection in SQL SELECT clause
+        /// </summary>
+        /// <param name="projection"> The projection expression. </param>
+        protected virtual void VisitProjection([NotNull] Expression projection) => Visit(ApplyOptimizations(projection, searchCondition: false));
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Rename these two to GenerateX, which is the pattern we are using for these kinds of virtual methods - Visit methods should deal with handling a specific node type.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 1 above)
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Yeah, was thinking its a good opportunity to rename it now as it doesn't have anything to do with JoinExpression

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Suggest a new name if you want change 😝

Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: how about ProcessExpressionList?

**Comment by smitpatel** [comment]
Updated.

**Pull request merged**